### PR TITLE
Add test methods that assert directory structure and content

### PR DESCRIPTION
**Pull request by mzganec**
Needed in https://github.com/genialis/resolwe-bio/pull/532.

### CODE REVIEW

**Comment by mstajdohar** [comment]
Rebase

File: resolwe/toolkit/tests/test_dirs.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        }
+
+        sym_link = {
+            'text file.txt': None,
+            'sym link': None,
+        }
+        upload_dir = self.run_process('upload-dir', {'src': 'compressed dir.tar.gz'})
+        self.assertDirExists(upload_dir, 'dir')
+        # https://github.com/PyCQA/pylint/issues/1653
+        self.assertRaisesRegex(  # pylint: disable=deprecated-method
```
Comment (dblenkus) [inline resolwe/toolkit/tests/test_dirs.py]: This is a test for `ProcessTestCase.assertDirExists`, not for the process.

File: resolwe/test/testcases/process.py [hunk 2]
```diff
@@ -535,6 +538,27 @@ def assertStatus(self, obj, status):  # pylint: disable=invalid-name
...
+        field_schema, field = None, None
+        for field_schema, field, field_path in iterate_fields(obj.output, obj.process.output_schema, ''):
+            if path == field_path:
+                break
+        else:
+            self.fail("Field not found in path {}.".format(path))
+
+        field_name = field_schema['name']
+        field_value = field[field_name]
+        return field_value, field_schema
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/test/testcases/process.py]: Changed to `return field[field_name], field_schema` so that we can still check field type in [assertDirExists](https://github.com/genialis/resolwe/pull/482/files#diff-9bac5f597553267222418cbf41d04daeR762).

File: resolwe/toolkit/tests/test_dirs.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+import os
+
+from resolwe.flow.models import Data
+from resolwe.test import ProcessTestCase, tag_process, with_docker_executor
+
+
+class DirsProcessTestCase(ProcessTestCase):
+
+    def setUp(self):
+        super(DirsProcessTestCase, self).setUp()
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/toolkit/tests/test_dirs.py]: Removed.

File: resolwe/test/testcases/process.py [hunk 4]
```diff
@@ -535,6 +538,27 @@ def assertStatus(self, obj, status):  # pylint: disable=invalid-name
...
         )
 
+    def _get_output_field(self, obj, path):
+        """Return object's output field value and schema.
+
+        :param obj: object with the output field
+        :type obj: ~resolwe.flow.models.Data
+
+        :param str path: path to
+            :class:`~resolwe.flow.models.Data` object's output field
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/test/testcases/process.py]: Done.

File: resolwe/test/testcases/process.py [hunk 5]
```diff
@@ -535,6 +538,27 @@ def assertStatus(self, obj, status):  # pylint: disable=invalid-name
...
+
+        """
+        field_schema, field = None, None
+        for field_schema, field, field_path in iterate_fields(obj.output, obj.process.output_schema, ''):
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/test/testcases/process.py]: This is an analog of iteration in [`assertFields`](https://github.com/genialis/resolwe/blob/master/resolwe/test/testcases/process.py#L555). If `path_prefix` is left set to `None`, we cannot make the comparison in the next line because in that case [`field_path` is not returned by `iterate_fields`](https://github.com/genialis/resolwe/blob/master/resolwe/flow/utils/iterators.py#L32).

File: resolwe/test/testcases/process.py [hunk 6]
```diff
@@ -729,6 +754,141 @@ def assertFilesExist(self, obj, field_path):  # pylint: disable=invalid-name
...
+        _, schema = self._get_output_field(obj, field_path)
+        if not schema['type'].startswith('basic:dir:'):
+            self.fail(msg='Field {} is not of type basic:dir:'.format(field_path))
+
+        field = dict_dot(obj.output, field_path)
+        dir_path = obj.location.get_path(filename=field['dir'])
+        if not os.path.isdir(dir_path):
+            self.fail(msg="Directory {} in output field {} does not exist.".format(dir_path, field_path))
+
+    def assertPathExists(self, obj, field_path):  # pylint: disable=invalid-name
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/test/testcases/process.py]: Removed.

File: resolwe/test/testcases/process.py [hunk 7]
```diff
@@ -729,6 +754,141 @@ def assertFilesExist(self, obj, field_path):  # pylint: disable=invalid-name
...
+        :param bool exact: if ``True`` tested directory structure must exactly match `dir_struct`.
+            If ``False`` `dir_struct` must be a partial structure of the directory to compare
+
+        """
+        self.assertDirExists(obj, field_path)
+        field = dict_dot(obj.output, field_path)
+        dir_path = obj.location.get_path(filename=field['dir'])
+        self._assert_dir_structure(dir_path, dir_struct, exact)
+
+    def _assert_dir(self, dir_path, fn, fail_on_funny=True):
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/test/testcases/process.py]: Renamed to `fn_correct` as in [`_assert_file`](https://github.com/genialis/resolwe/blob/master/resolwe/test/testcases/process.py#L587).

File: resolwe/toolkit/tests/test_dirs.py (hunk 1 above)
Comment (mzganec) [reply resolwe/toolkit/tests/test_dirs.py]: Removed.

File: resolwe/test/tests/test_testing_framework.py [hunk 8]
```diff
@@ -195,6 +197,123 @@ def test_assert_almost_equal_generic(self):  # pylint: disable=invalid-name
...
+            open(os.path.join(tmpdirname, 'text file.txt'), 'w').close()
+            os.mkdir(os.path.join(tmpdirname, 'dir'))
+            open(os.path.join(tmpdirname, 'dir', 'another file'), 'w').close()
+            os.mkdir(os.path.join(tmpdirname, 'dir', 'empty dir.ext'))
+
+            # Exact structure checks
+            ProcessTestCase._assert_dir_structure(  # pylint: disable=protected-access
+                self.dummy_case, tmpdirname, correct_structure)
+
+            # https://github.com/PyCQA/pylint/issues/1653
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/test/tests/test_testing_framework.py]: Added to class `setUp` method.

File: resolwe/test/tests/test_testing_framework.py [hunk 9]
```diff
@@ -195,6 +197,123 @@ def test_assert_almost_equal_generic(self):  # pylint: disable=invalid-name
...
+            AssertionError, 'Directory .* does not exist.', ProcessTestCase.assertDir,
+            self.dummy_case, obj_mock, 'field_name', 'foo.tar.gz')
+
+    def test_assert_dir_compressed_missing(self):  # pylint: disable=invalid-name
+        test_output = io.BytesIO()
+        correct_output = io.BytesIO()
+        open_mock = mock.MagicMock(return_value=tarfile.open(fileobj=test_output, mode='w:gz'))
+
+        with tempfile.TemporaryDirectory() as tmpdirname:
+            open(os.path.join(tmpdirname, 'extensionless file'), 'w').close()
```
[the remark this answers is not in the document]
Comment (mzganec) [reply resolwe/test/tests/test_testing_framework.py]: Added to class `setUp` method and prepared the corresponding `tearDown` method.

**Comment by mstajdohar** [comment]
@dblenkus Please do another review pass.

**Comment by mstajdohar** [comment]
@mzganec Please rebase

**Comment by mstajdohar** [comment]
@dblenkus ping

File: resolwe/test/testcases/process.py [hunk 10]
```diff
@@ -535,6 +538,24 @@ def assertStatus(self, obj, status):  # pylint: disable=invalid-name
...
+        """Return object's output field value and schema.
+
+        :param obj: object with the output field
+        :type obj: ~resolwe.flow.models.Data
+
+        :param str path: path to :class:`~resolwe.flow.models.Data`
+            object's output field
+
+        """
+        field_schema, field = None, None
```
Comment (dblenkus) [inline resolwe/test/testcases/process.py]: This is not needed.

File: resolwe/test/testcases/process.py [hunk 11]
```diff
@@ -535,6 +538,24 @@ def assertStatus(self, obj, status):  # pylint: disable=invalid-name
...
+        for field_schema, field, field_path in iterate_fields(obj.output, obj.process.output_schema, ''):
+            if path == field_path:
+                field_name = field_schema['name']
+                return field[field_name], field_schema
```
Comment (dblenkus) [inline resolwe/test/testcases/process.py]: Where do you need `field[field_name]`?

File: resolwe/test/testcases/process.py [hunk 12]
```diff
@@ -729,6 +750,127 @@ def assertFilesExist(self, obj, field_path):  # pylint: disable=invalid-name
...
+        :param obj: object that includes the file for which to check if
+            it exists
+
+        :param field_path: directory name/path
+        """
+        _, schema = self._get_output_field(obj, field_path)
+        if not schema['type'].startswith('basic:dir:'):
+            self.fail(msg='Field {} is not of type basic:dir:'.format(field_path))
+
+        field = dict_dot(obj.output, field_path)
```
Comment (dblenkus) [inline resolwe/test/testcases/process.py]: You can return `field` from `_get_output_field`.

File: resolwe/test/testcases/process.py [hunk 13]
```diff
@@ -729,6 +750,127 @@ def assertFilesExist(self, obj, field_path):  # pylint: disable=invalid-name
...
+            match `dir_struct`. If ``False`` `dir_struct` must be a partial
+            structure of the directory to compare
+
+        """
+        self.assertDirExists(obj, field_path)
+        field = dict_dot(obj.output, field_path)
+        dir_path = obj.location.get_path(filename=field['dir'])
+        self._assert_dir_structure(dir_path, dir_struct, exact)
+
+    def _assert_dir(self, dir_path, fn_correct, fail_on_funny=True):
```
Comment (dblenkus) [inline resolwe/test/testcases/process.py]: Is `fail_on_funny` still needed?

File: resolwe/test/testcases/process.py [hunk 14]
```diff
@@ -729,6 +750,127 @@ def assertFilesExist(self, obj, field_path):  # pylint: disable=invalid-name
...
+        :param str fn: file name (and relative path) of the correct compressed
+            directory to compare against. Path should be relative to the
+            ``tests/files`` directory of a Django application. Compressed
+            directory needs to be in ``tar.gz`` format.
+
+        """
+        self.assertDirExists(obj, field_path)
+        field = dict_dot(obj.output, field_path)
+        dir_path = obj.location.get_path(filename=field['dir'])
+        self._assert_dir(dir_path, fn)
```
Comment (dblenkus) [inline resolwe/test/testcases/process.py]: Is there any advantage of moving this to a separate method?

File: resolwe/toolkit/processes/dirs.yml [hunk 15]
```diff
@@ -0,0 +1,36 @@
+# ===================
```
Comment (dblenkus) [inline resolwe/toolkit/processes/dirs.yml]: This should go into a separate commit.

Comment (dblenkus) [reply resolwe/toolkit/processes/dirs.yml]: And don't forget to add a change log.

File: resolwe/toolkit/tests/test_dirs.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+
+class DirsProcessTestCase(ProcessTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.files_path = os.path.abspath(os.path.join(os.path.dirname(__file__), 'files'))
+
+        test_case = unittest.TestCase()
+        self.assertRaisesRegex = test_case.assertRaisesRegex  # pylint: disable=invalid-name
```
Comment (dblenkus) [inline resolwe/toolkit/tests/test_dirs.py]: This can be removed as we upgraded to Django 2.

File: resolwe/toolkit/tests/test_dirs.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+
+    @with_docker_executor
+    @tag_process('upload-dir')
+    def test_upload_dir(self):
```
Comment (dblenkus) [inline resolwe/toolkit/tests/test_dirs.py]: This test tests too many things. You just need to make sure that process works as expected, so chose only one assertion function. All other cases should be tested in `test_assert_dir_structure`, where assertion functions are tested.

File: resolwe/test/testcases/process.py (hunk 13 above)
Comment (mzganec) [reply resolwe/test/testcases/process.py]: I would leave it in in case we encounter something nasty, but we don't have a specific use-case at the moment.

File: resolwe/test/testcases/process.py (hunk 14 above)
Comment (mzganec) [reply resolwe/test/testcases/process.py]: I followed the same design as in the [file counterpart](https://github.com/genialis/resolwe/blob/master/resolwe/test/testcases/process.py#L587). I presume that this is needed to be able to [test this method](https://github.com/genialis/resolwe/blob/ca9c22bcc057accbb8d8999722e289c6664ed09b/resolwe/test/tests/test_testing_framework.py#L211) without explicitly running processes.
